feat: add spec for namespace configs

hperl opened the pull request:
This PR adds the first version of the Ory Permission language spec to the repo.

Two (last) open issues:

- [x] Type unions
- [ ] Type representation of subject sets

zepatrik commented on docs/namespace_config_spec.md:
````diff
@@ -0,0 +1,89 @@
+Identifiers name program entities such as variables and types. An identifier is a sequence of one or more letters and digits. The first character in an identifier must be a letter.
+
+```ebnf
+identifier = letter { letter | unicode_digit } .
````
We need to define letters here, I assume unicode? What about `-` and `_`?

kevgo reviewed (commented):
Thanks for writing this up! This is really cool, I'm very excited to start playing with this idea more! I consider this idea of using TS syntax pre-alpha stage, so here comes relatively fundamental feedback. Curious what you think!

hperl commented on go.mod:
```diff
@@ -8,11 +8,7 @@ replace (
-	github.com/dgrijalva/jwt-go => github.com/golang-jwt/jwt/v4 v4.0.0
-	github.com/opencontainers/runc => github.com/opencontainers/runc v1.0.3
-)
+replace github.com/dgrijalva/jwt-go => github.com/golang-jwt/jwt/v4 v4.0.0
```
This was cherry-picked on a PR that already got merged, should disappear with a rebase ;)

hperl commented on docs/namespace_config_spec.md:
```diff
@@ -0,0 +1,281 @@
+* `permissions`
+* `this`
+* `ctx`
+* `id`
```
Eventually these will be added. None are part of the language right now, and I'd like to implement the simplest thing first and then expand from there.

hperl commented on docs/namespace_config_spec.md:
```diff
@@ -0,0 +1,281 @@
+  }
+
+  permissions = {
+    isViewer: (ctx: Context) => this.relations.viewers.includes(ctx.subject),
```
Good catch! I think we should pick (b) and take it out of the examples. Probably real helper methods would then be methods directly on the class, or functions as you suggested in (c).

hperl commented on docs/namespace_config_spec.md:
````diff
@@ -0,0 +1,281 @@
+}
+```
+
+Then the config reads as follows (note that `metadata` is now mandatory, and
+that only the specified array methods can be used, etc.):
````
What I meant to say is you get a type error in TypeScript if you miss `metadata`, since it then no longer implements `Namespace`.

hperl commented on docs/namespace_config_spec.md:
```diff
@@ -0,0 +1,281 @@
+}
+
+interface Array<Namespace> {
+  includes(searchElement: Namespace): boolean;
```
Right!

hperl commented on docs/namespace_config_spec.md:
```diff
@@ -0,0 +1,281 @@
+
+## Examples
+
+The config can be typechecked by TypeScript with the `noLib` option (preventing
+the standard globals) and supplying the following globals instead:
```
I clarified this. Regarding tooling, we will probably ship a TypeScript config that works best for writing the Keto config. I just need to research the best way to do it.

hperl commented on docs/namespace_config_spec.md:
````diff
@@ -0,0 +1,281 @@
+```
+* a `TranstitiveCheck`, a call to a permission on a relation, e.g., `this.relations.parents.some(p => p.permissions.view(ctx))`:
+```ebnf
+TransitiveCheck = "this" "." "relations" "." RelationName "." "some" "(" Var "=>" PermissionCall ")" .
````
We decided to leave this out of the first version, since that semantic does not translate to Zanzibar concepts yet (Zanzibar has no set intersection for rewrites). For top-level intersection / every / and, you can use `&&`.

hperl commented on docs/namespace_config_spec.md:
````diff
@@ -0,0 +1,281 @@
+
+```ebnf
+PermissionBody  = PermissionCheck | { Operator PermissionCheck } .
+Operator        = "||" | "&&" .
````
Negation does not map to Zanzibar concepts. We can extend them later and add it to the spec.

hperl commented on docs/namespace_config_spec.md:
```diff
@@ -0,0 +1,281 @@
+### Comments
+
+1. Line comments start with the character sequence // and stop at the end of the line.
+2. General comments start with the character sequence /* and stop with the first subsequent character sequence */.
```
Yes, good idea!

hperl commented on docs/namespace_config_spec.md:
```diff
@@ -0,0 +1,281 @@
+* `Array<Namespace>` or `Namespace[]`
+* boolean
+
+In TypeScript, they would be defined as follows:
```
Keto Schema Language?

hperl commented on docs/namespace_config_spec.md:
```diff
@@ -0,0 +1,281 @@
+| `(`, `)` | `x()` | call function `x` |
+| `[`, `]` | `T[]` | `T` is an array type |
+| `{`, `}` | `class x {}` | Scope delimiters |
+| `=>` | `list.some(el => f(el))` | Lambda definition token. |
```
Yes, let's leave it out of v1 of the spec.

hperl commented on docs/namespace_config_spec.md:
```diff
@@ -0,0 +1,281 @@
+
+### Comments
+
+1. Line comments start with the character sequence // and stop at the end of the line.
```
It was actually copied from the Go spec. I think we could reference many specs, as this is the commenting scheme for many langauges.

hperl commented on docs/namespace_config_spec.md:
```diff
@@ -0,0 +1,281 @@
+* `Context`
+* `Namespace`
+* `Array<Namespace>` or `Namespace[]`
+* boolean
```
Good catch!

hperl commented on docs/namespace_config_spec.md:
```diff
@@ -0,0 +1,281 @@
+
+The Keto namespace configuration language is a syntactical subset of TypeScript.
+Along with type definitions for the syntax elements of the language (such as
+`Namespace` or `Context`), the user can get context help from her IDE while
```
Added, thanks!

hperl commented:
@kevgo thank you for your comments. I addressed your suggestions, and especially liked the improved wordings for `related` and `permits`.

Some of the things you mentioned (adding keywords, builtins), can be moved to a next version of the spec, as they will be backwards-compatible.

The only thing that is still open is the discussion about `==` vs `===`. I find comparison irritating in JS as well and would rather define simple semantics in our language. If we just pick `===`, then it feels strange to not have `==`. Having both (with same or different semantics) would be confusing as well. I would argue for just having `==` and forbidding `===`.

@aeneasr @zepatrik WDYT?

aeneasr commented on docs/namespace_config_spec.md:
```diff
@@ -0,0 +1,309 @@
+# The "Keto Schema Language" Specification
```
See comment below

aeneasr commented on docs/namespace_config_spec.md:
```diff
@@ -0,0 +1,309 @@
+chose to define our permissions configuration language as a subset of the most
+common general-purpose programming language: JavaScript/TypeScript.
+
+The Keto Schema Language is a syntactical subset of TypeScript. Along with type
```
I would almost prefer to give this a unique name, like "Ory Session Token", or "Ory Identity Schema":

```suggestion
The Ory Permission Schema Language is a syntactical subset of TypeScript. Along with type
```

aeneasr commented on docs/namespace_config_spec.md:
````diff
@@ -0,0 +1,309 @@
+The syntax is specified using Extended Backus-Naur Form (EBNF):
+
+```ebnf
+Production  = production_name "=" [ Expression ] "." .
````
Is `Production` common practice wording?

hperl replied:
Yes, this section is borrowed from the Go language spec.

hperl replied:
We could just remove this section completely and reference the Go spec. WDYT?

aeneasr replied:
In that case I think this is fine :) This probably will never change!

aeneasr commented on docs/namespace_config_spec.md:
````diff
@@ -0,0 +1,309 @@
+```
+
+Lower-case production names are used to identify lexical tokens. Non-terminals
+are in CamelCase. Lexical tokens are enclosed in double quotes "".
````
```suggestion
are in CamelCase. Lexical tokens are enclosed in double quotes `""`.
```

aeneasr commented on docs/namespace_config_spec.md:
````diff
@@ -0,0 +1,309 @@
+```ebnf
+identifier = letter { letter | digit } .
+digit      = "0" … "9"
+letter     = "A" … "Z" | "a" … "z" | "-" | "_"
````
Are identifiers things like variables? If so, can we really use `-`?

```
// This would not compile in TS:
var foo-bar
```

kevgo replied:
Good point, let's stick with the [JS syntax for identifiers](https://developer.mozilla.org/en-US/docs/Glossary/Identifier): unicode letters, $, _, and digits (0-9).

hperl replied:
Yes, I noticed that as well. Thanks!

aeneasr commented on docs/namespace_config_spec.md:
```diff
@@ -0,0 +1,281 @@
+| --- | --- | --- |
+| `&&` | *x* `&&` *y* | true iff. *x* and *y* are true |
+| `\|\|` | *x* `\|\|` *y* | true iff. *x* or *y* are true |
+| `==` | *x* `==` *y* | true iff. *x* equals *y* |
+| `!=` | *x* `!=` *y* | ture iff. *x* does not equal *y* |
```
I think kevgo has a point here. Novice JS users will probably not care about this, but experienced devs will for sure ask themselves what this means and might also be confused as to why `===` is not working. We should also explain what `equal` means here?

kevgo replied:
I agree that `===` is clunky and the whole situation is far from ideal. Here are a few things that I think we should keep in mind. 

The `==` operator has [very well defined behavior](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Equality) in JS. If we use `==` in our language, we have to remove the parts from its spec where it tries to convert the operands into types that are comparable. Technically, this is within our mission of definining a "subset" of TypeScript. 

I am concerned that removing elements from the spec of `==` effectively changes its behavior. This will confuse humans who likely expect a subset where the parts left in the language are "normal" JS. The TS tooling we plan on using (type checker, linter, unit tests running in Node) will always assume the JS behavior. I want to avoid situations where the behavior of a Keto spec is different in tests than when running inside Keto. So I guess I propose that we at least test this before finalizing this spec.

For reference, the `===` operator, while uglier, [already does what we want](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Strict_equality). Unless "equality" means something completely different in Keto than it does in JS. Is that the case here?

hperl replied:
OK this makes sense. Then we will *only* support the `===` and `!==` operators, as this captures best the semantics of the rules in intersection with the semantics of JS.

hperl replied:
On second thought, since we right now in the spec only allow relations with type `T[]`, checks for equality are actually done implicitly via `includes`. This means we can take direct equality out of the spec for now.

aeneasr replied:
> OK this makes sense. Then we will _only_ support the `===` and `!==` operators, as this captures best the semantics of the rules in intersection with the semantics of JS.

I agree with this

> On second thought, since we right now in the spec only allow relations with type `T[]`, checks for equality are actually done implicitly via `includes`. This means we can take direct equality out of the spec for now.

True, none of the examples actually use an equality operator. Are there use cases where this is needed actually?

kevgo commented on docs/namespace_config_spec.md:
````diff
@@ -0,0 +1,309 @@
+instead:
+
+```typescript
+type Context = { subject: never }
+
+interface Namespace {
+    metadata: { id: number }
+    related?: { [relation: string]: Namespace[] }
+    permits?: { [method: string]: (ctx: Context) => boolean }
+}
+
+interface Array<Namespace> {
+  includes(element: Namespace): boolean;
+  some(callbackfn: (value: Namespace) => boolean): boolean;
+}
````
This part is already listed in identical form on lines 105-116. Can we remove one of them?

hperl replied:
yes

kevgo reviewed (commented):
Awesome collaboration on this, thanks for all the thoughful comments! I totally agree with keeping this as simple as possible. Another round of feedback after sleeping over it again. This is coming along nicely!

hperl commented on docs/namespace_config_spec.md:
```diff
@@ -0,0 +1,309 @@
+
+* `Context`
+* `Namespace`
+* `Array<Namespace>` or `Namespace[]`
```
Agreed!

hperl commented on docs/namespace_config_spec.md:
```diff
@@ -0,0 +1,309 @@
+type Context = { subject: never }
+
+interface Namespace {
+    metadata: { id: number }
```
That makes sense, but will trigger some internal work in creating the mapping table you mentioned. @zepatrik FYI.

hperl commented:
> Awesome collaboration on this, thanks for all the thoughful comments! I totally agree with keeping this as simple as possible. Another round of feedback after sleeping over it again. This is coming along nicely!

Agreed! I just incorporated the latest comments and am quite happy with the result. I noticed that we can sidestep the equality discussion because we are only testing for set membership for now anyways (and there are good arguments for it staying that way).

kevgo reviewed (commented):
The next round of feedback. I limit the amount of ideas in each review to keep them manageable. What do you think?

aeneasr commented on docs/namespace_config_spec.md:
```diff
@@ -0,0 +1,319 @@
+# The "Ory Permission Language" Specification
```
Naming wise it is important to have Ory in it. Keto might be renamed to Ory Permission Service at some point and it should fit in the other terminology we have (Ory Session, Ory Identity Schema, …)

aeneasr replied:
Let's go with Ory Permission Language. Noone really writes OPL any more, it should be fine

kevgo replied:
Good points. IBM also has an [Optimization Programming Language](https://www.ibm.com/docs/en/icos/20.1.0?topic=opl-optimization-programming-language) that still seems active. Maybe we go with "Ory Permission Format" (OPF) or "Ory Permission Grammar" (OPG)? That's also better googleable.

We chose TypeScript mostly because it's popular. In the future, we might support additional programming languages since not everybody likes, wants, or uses TypeScript (and the language has so many warts). So, maybe we encode the language used in the name of the config format?

- OP-TS or OPTS: Ory Permissions as TypeScript
- OP-GO or OPGO: Ory Permissions as Go
- OP-RS or OPRS: Ory Permissions as Rust
- OP-RB or OPRB: Ory Permissions as Ruby
- OP-PY or OPPY: Ory Permissions as Python

zepatrik commented on docs/namespace_config_spec.md:
```diff
@@ -0,0 +1,319 @@
+Note that all relations are defined as array types `T[]`, because there are
+naturally only many-to-many relations in Keto.
+
+The following declares a type *Document* with three relations: *owners* and
+*viewers*, both of which have *users* as subjects. Additionally, the relation
```
It is the other way around in your example. This is one huge topic that everyone gets wrong at some point. That's why we think the object-oriented style of this schema language will help, but apparently there are still confusions :sweat_smile:

kevgo replied:
@zepatrik I don't follow, can you please elaborate which of my two examples you refer to and what you see as the right way of reading this? TBH I can think of a whole number of ways of reading this. My examples assume this order in the document:

```ts
class Document { permits: { view: (ctx: Context) => this.owners.includes(ctx.subject) } }
```

1. right-middle-left: "Only owners can view documents"
2. left-middle-right: "Documents can only be viewed by their owners"
3. middle-left-right: "You can only view documents that you own."

If this is a huge topic that everyone gets wrong at some point, this document should define the idiomatic way of reading rules in our format and provide guidance and examples.

I also think if we can do better than the Google recommendation of reading right-to-left, that's another big advantage for us. 

I find (3) the most intuitive, then (2), then (1). I like that (2) follows the natural reading direction and the way I think about OO code. Hence my recommendation is for (2). What do you think?

kevgo replied:
The spec currently refers to "subjects" (on line 169 and 208) and "objects" (on line 168) of relations. I think the spec should either define these terms or remove them.

hperl commented on docs/namespace_config_spec.md:
````diff
@@ -0,0 +1,319 @@
+
+```ebnf
+string_lit    = single_quoted | double_quoted .
+single_quoted = "'" { letter | digit } "'" .
````
Typescript has single quotes as well; and from the implementation aspect the effort is negligible.

kevgo replied:
Please keep in mind that both string types have slightly different parsing rules. For example, double-quoted strings don't require escaping single quotes and it's the opposite for single-quoted strings. 

If the goal is to support all string types that TypeScript supports, we might also want to support [template strings](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals) (strings in backticks). This string type supports interpolation and multilines.

hperl commented on docs/namespace_config_spec.md:
```diff
@@ -0,0 +1,319 @@
+    permits?: { [method: string]: (ctx: Context) => boolean }
+}
+
+interface Array<Namespace> {
```
The type definition is not part of the configuration language, it is just part of the tooling. We never have to parse the type definitions. Therefore, we can still use the `Array<...>` syntax here, since this is part of TypeScript.

hperl commented on docs/namespace_config_spec.md:
```diff
@@ -0,0 +1,319 @@
+
+* `class`
+* `implements`
+* `metadata`
+* `related`
+* `permits`
```
I don't think it makes much difference for the user, but for the parser I found it convenient to have these as keywords. Removing keywords is always backwards-compatible, so we can demote e.g. `metadata` to an identifier later.

hperl commented on docs/namespace_config_spec.md:
```diff
@@ -0,0 +1,319 @@
+}
+
+interface Array<Namespace> {
+  includes(searchElement: Namespace): boolean;
```
I don't care much about the spelling, but I took it from the MDN docs and the TypeScript type implementation, see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/includes.

hperl commented on docs/namespace_config_spec.md:
```diff
@@ -0,0 +1,319 @@
+    id: "2"
+  }
+  related = {
+    parent: Document[]
```
Yes, this is more correct. We can't specify that we want only one, or exactly one parent in Keto, since there are always just relation tuples in the database (at least these constraints would involve fundamental changes to the data store).

We should encourage users to always pluralize their relations, as already suggested by the now-mandatory array type.

hperl commented on docs/namespace_config_spec.md:
```diff
@@ -0,0 +1,319 @@
+### Relation declaration
+
+Inside the class declaration, relations are declared. `RelationName` must be
+unique identifiers and are used as the relation strings in the tuples. The
```
Here, `RelationName` refers to "owners", not "core.User[]". The relations in Keto are global identifiers, meaning they have no namespace.

Namespaces can be scoped as you suggested, but I would like to keep a module system out of the first version of the spec.

hperl commented on docs/namespace_config_spec.md:
```diff
@@ -0,0 +1,319 @@
+
+| Operator | Signature | Semantic |
+| --- | --- | --- |
+| `&&` | *x* `&&` *y* | true iff. *x* and *y* are true |
+| `\|\|` | *x* `\|\|` *y* | true iff. *x* or *y* are true |
```
Agreed. Currently the language only allows these operators in very few places (when comapring bools).

kevgo replied:
What about my suggestion from above, which "both" and "either" to make the difference between logical boolean operators from the common usage in English more clear?

hperl commented on docs/namespace_config_spec.md:
```diff
@@ -0,0 +1,319 @@
+| Operator | Example | Semantic |
+| --- | --- | --- |
+| `(`, `)` | `x()` | call function `x` |
+| `[`, `]` | `T[]` | `T` is an array type |
```
Yes, thanks!

hperl reviewed (commented):
## TODO
- remove metadata
- parens in boolean expr

hperl reviewed (commented):
address review comments

kevgo reviewed (commented):
Excellent, that's all the feedback I have from my end. Next steps after the last changes of the spec are done:

- polish of the language using grammarly.com (I can do this, want to wait after all other edits are done since this will cause lots of conflicts with concurrent edits)
- format this file using Prettier

zepatrik reviewed (approved):
Very nice :+1:

kevgo commented on docs/ory_permission_language_spec.md:
```diff
@@ -0,0 +1,298 @@
+technology solutions that protect privacy and identity in the information age. A
+number of proprietary languages to represent permission already exists, namely
+the [Authzed Schema Language](https://docs.authzed.com/reference/schema-lang)
+and [Auth0 FGA](https://docs.fga.dev/). Many permissions are defined by normal
```
```suggestion
and [Auth0 FGA](https://docs.fga.dev/). Most permissions are defined by normal
```

kevgo commented on docs/ory_permission_language_spec.md:
```diff
@@ -0,0 +1,298 @@
+developers who typically are most familiar with web technologies like JavaScript
+or Typescript. We therefore see the need for a developer-friendly configuration
+language for permissions that has such a small learning curve that it can be
+picked up, used, and understood with confidence by most developers. We therefore
```
```suggestion
picked up, used, and understood with confidence by most developers. We
```

kevgo commented on docs/ory_permission_language_spec.md:
```diff
@@ -0,0 +1,298 @@
+
+### Type declaration
+
+The top level of the configuration consists of a list of `class` declaration for
```
```suggestion
The top level of the configuration consists of a list of `class` declarations for
```

kevgo commented on docs/ory_permission_language_spec.md:
````diff
@@ -0,0 +1,298 @@
+ClassSpec       = [ RelationDecls ] | [ PermissionDefns] .
+```
+
+The following example declares a type _User_ that is never the object in any
+relation. It can, however, be the subject of a relation, the relation being
+declared on some other type.
````
I think leaving out this object-subject stuff makes this more clear.

```suggestion
The following example declares a type _User_.
```

kevgo commented on docs/ory_permission_language_spec.md:
```diff
@@ -0,0 +1,298 @@
+
+### Relation declaration
+
+Inside the class declaration, relations are declared. `RelationName` must be
```
```suggestion
Relations are defined inside the class declaration. `RelationName` must be
```

kevgo commented on docs/ory_permission_language_spec.md:
```diff
@@ -0,0 +1,298 @@
+### Relation declaration
+
+Inside the class declaration, relations are declared. `RelationName` must be
+unique identifiers and are used as the relation strings in the tuples. The
```
```suggestion
unique identifiers used as the relation strings in the tuples. The
```

kevgo commented on docs/ory_permission_language_spec.md:
```diff
@@ -0,0 +1,298 @@
+
+Inside the class declaration, relations are declared. `RelationName` must be
+unique identifiers and are used as the relation strings in the tuples. The
+`TypeName` must be the name of a `type` that is defined above or below (in
```
```suggestion
`TypeName` must be the name of an already defined `type` (in
```

kevgo commented on docs/ory_permission_language_spec.md:
````diff
@@ -0,0 +1,298 @@
+PermissionCheck = TransitiveCheck | IncludesCheck .
+```
+
+The body of a permission check is either one of
````
```suggestion
The body of a permission check is either one of:
```

kevgo commented on docs/ory_permission_language_spec.md:
```diff
@@ -0,0 +1,298 @@
+
+#### Implementation note
+
+`IncludeCheck` and `TransitiveCheck` will be translated to Zanzibar concepts as
```
```suggestion
`IncludeCheck` and `TransitiveCheck` translate to Zanzibar concepts as
```

kevgo reviewed (commented):
Here are the suggestions from grammarly.com. Mostly removing passive voice. Please run Prettier again after you add these changes.

hperl commented:
> Excellent, that's all the feedback I have from my end. Next steps after the last changes of the spec are done:
> 
> * polish of the language using grammarly.com (I can do this, want to wait after all other edits are done since this will cause lots of conflicts with concurrent edits)
> * format this file using Prettier

Thanks, @kevgo! I (think) I applied all of your suggestions.

kevgo commented on docs/ory_permission_language_spec.md:
````diff
@@ -0,0 +1,303 @@
+
+```ebnf
+string_lit    = single_quoted | double_quoted .
+single_quoted = "'" { letter | digit } "'" .
+double_quoted = '"' { letter | digit } '"' .
````
Can strings contain spaces too?

hperl replied:
Ah, now that we removed metadata we don't have to have this in the spec at all!

kevgo commented on docs/ory_permission_language_spec.md:
```diff
@@ -0,0 +1,303 @@
+strings in the tuples. The `TypeName` must be the name of another `type` that is
+defined above or below (in TypeScript: a class that implements `Namespace`).
+
+Type unions (`|`) can be used to denote that a relation can have subjects of
+multiple types, e.g., `viewers: User[] | Group[]`.
```
Wow this is cool!

kevgo reviewed (approved):
This has turned out really nice! Can't wait to build some permission sets in it!
